### PR TITLE
Fix transitive link dependencies for contrib ports

### PR DESCRIPTION
**Pull request by ypujante**
This is the PR for issue #21597

After digging in the code, I realized that old ports were using the `process_dependencies` concept to add link time dependencies. For example, compiling code with port `sdl2_mixer` does **not** require to also add `sdl2` or other dependencies manually for the linking process to work. This is because  `process_dependencies` does the "right" thing (but it uses `settings` which is not what new port use)

So this PR allow ports that don't use `settings` to implement a similar functionality by simply returning their list of link time dependencies.

In my use case:

```python
# from imgui.py contrib port
def process_dependencies(settings):
    return ['contrib.glfw3']
```

Because `process_dependencies` currently does not return anything, it is 100% backward compatible. This change does not affect any current port.

If you are ok with this PR I will obviously update the documentation to explain the change.

Note, if you want more details, I have this [other PR](https://github.com/emscripten-core/emscripten/pull/21601) that I have now closed which has more details about the reasoning

### CODE REVIEW

File: tools/ports/__init__.py
Comment (sbc100) [inline tools/ports/__init__.py:398]: I think just `if port_dependencies:` is fine here.

Comment (sbc100) [reply tools/ports/__init__.py:398]: If you look at sdl2_image, it does two things during `process_dependencies`.

1. Update settings
2. Update its own `.deps` list

I have two siggestions here:

1. Can we stick to the mechanism of updated `.deps` during `process_dependencies` rather than returning the deps?
2. I think `resolve_dependencies` is supposed to add the dependencies to `port_set`,  I think it would then be up to the calling code to add them to `ports_needed`.

Comment (ypujante) [reply tools/ports/__init__.py:398]: I am not sure I fully understand. Are you suggesting that I should do this in my code (and no change in emscripten at all):

```python
# imgui.py port
def process_dependencies(settings):
    ports_needed.add('contrib.glfw3')
```

If I can get a hold on `ports_needed` then clearly that could work.

Comment (sbc100) [reply tools/ports/__init__.py:398]: No, I believe you your dependencies are fixed/static then you shouldn't need to implement `process_dependencies` at all. 

It should be enough to just write `deps = ['contrib.glfw3']`

`process_dependencies` should only be needed for ports that have dyanmic/conditional dependencies.

Comment (ypujante) [reply tools/ports/__init__.py:398]: So how do I fix the link issue then?

With "old" port there is no link issue. With new ports there is one. I don't think that makes sense.

```text
# this is fine and will automatically link with sdl2 even if not explicitly added to the command line
emcc --use-port=sdl2_mixer test/hello_world.c -o /tmp/out.js
```

Why would it work for "old" port but not new ones?

Comment (ypujante) [reply tools/ports/__init__.py:398]: You said:

> I think resolve_dependencies is supposed to add the dependencies to port_set, I think it would then be up to the calling code to add them to ports_needed.

I don't understand what is the calling code you are referring to that would "add them to ports_needed"

Comment (sbc100) [reply tools/ports/__init__.py:398]: I think the problem is that we don't add the deps to the new `ports_needed` global.

We have a `if port.needed(settings):` check on line 500 which we should perhaps delete.

Alternatively you could make `port.needed(settings)` return true if you add  `ports_needed.add(dep)` right after `port_set.add(deps)` perhaps?

I think perhaps we should do both of the above since the check on line 500 seems redundant, but it would also be nice if `.needed()` returned the correct thing for contrib deps

Comment (ypujante) [reply tools/ports/__init__.py:398]: Ok I will investigate further

In regard to the statement 

> process_dependencies should only be needed for ports that have dyanamic/conditional dependencies.

I believe it is not entirely accurate. You can check [vorbis](https://github.com/emscripten-core/emscripten/blob/main/tools/ports/vorbis.py) port:

```python
def process_dependencies(settings):
  settings.USE_OGG = 1
```

There is nothing dynamic here. This function lets you write `emcc -sUSE_VORBIS=1 ...` without having to write `emcc -sUSE_OGG=1 -sUSE_VORBIS=1 ...` for linking purposes...

Comment (sbc100) [reply tools/ports/__init__.py:398]: Ah yes, I see what you mean.  I guess we have a lot of duplication between `process_dependencies` and `deps`. 

I would hope that we could simply rely on `deps` along rather then duplication between them. e.g.:

```
deps = ['freetype', 'sdl2', 'harfbuzz']  

...

def process_dependencies(settings):                                                                  
  settings.USE_SDL = 2                                                                               
  settings.USE_FREETYPE = 1                                                                          
  settings.USE_HARFBUZZ = 1    
```

At least for contrib ports it should be easy enough to just use `deps` I think.. since you don't have any settings that you need to set.

**Comment by ypujante** [comment]
I reverted the changes and went for the simple solution of removing the test entirely on line 500 as you suggested.

I am not entirely sure about what you mean with the alternate solution but would be open to implement it if you provide further details:

> Alternatively you could make port.needed(settings) return true if you add ports_needed.add(dep) right after port_set.add(deps) perhaps?

**Comment by ypujante** [comment]
In regards to failing tests:

* `test-windows`: test_minimal_runtime_code_size_hello_webgl2_wasm,  test_minimal_runtime_code_size_hello_webgl2_wasm2js, test_minimal_runtime_code_size_hello_webgl_wasm, test_minimal_runtime_code_size_hello_webgl_wasm2js
* `test-other`: same 4 failures
* `test-mac-arm64`: same 4 failures
* `test-mac`: same 4 failures

I run these tests locally on this branch and they fail too. I then switched to latest and seeing the same:

```text
# git rev-parse HEAD
0ca463011d0db12de9954b6b15d03f88dcd1eeaa
# ./test/runner other.test_minimal_runtime_code_size_hello_webgl_wasm2js
.... truncated
size of a.js == 21783, expected 21784, delta=-1 (-0.00%)
size of a.js.gz == 11439, expected 11435, delta=4 (+0.03%)
Total output size=22137 bytes, expected total size=22138, delta=-1 (-0.00%)
Total output size gzipped=11705 bytes, expected total size gzipped=11701, delta=4 (+0.03%)
Hey amazing, overall generated code size was improved by 1 bytes!
If this is expected, rerun the test with --rebaseline to update the expected sizes
FAIL

======================================================================
FAIL: test_minimal_runtime_code_size_hello_webgl_wasm2js (test_other.other)
```

File: tools/ports/__init__.py
Comment (sbc100) [inline tools/ports/__init__.py:404]: Could you add this line after ilne 500 instead?

**Review by sbc100 — COMMENTED** [review]
The code size test failure occur from time to time due to llvm changes.  

I will try to rebaseline them today, but you can generally ignore them in this case of change since they are unrelated.

File: tools/ports/__init__.py
Comment (ypujante) [reply tools/ports/__init__.py:404]: I don't understand. Line 500 is this:

```python
  for port in dependency_order(needed):
    port.linker_setup(Ports, settings)
    # port.get returns a list of libraries to link
    ret += port.get(Ports, settings, shared)
```

And your original suggestion was: _add ports_needed.add(dep) right after port_set.add(deps) perhaps?_

Comment (ypujante) [reply tools/ports/__init__.py]: For the record:

* I only made this change because you suggested it although I thought it was not necessary
> We have a if port.needed(settings): check on line 500 which we should perhaps delete.
> **Alternatively you could make port.needed(settings) return true if you add ports_needed.add(dep) right after port_set.add(deps) perhaps?**
> I think perhaps **we should do both** of the above since the check on line 500 seems redundant, but it **would also be nice** if .needed() returned the correct thing for contrib deps
* The call `node.process_dependencies(settings)` on line 397 is potentially modifying dependencies (see `sdl2_mixer.py` for an example), so unfortunately calling `resolve_dependencies` is not without side effects and modify dependencies. Every time it gets called, the `deps` field of the port gets modified and gets longer and longer... So I don't think that `resolve_dependencies` should casually be called to "simply to retrieve the dependency list"

I have now removed this line. So the fix for this issue is to simply remove the test at line 500. Which also makes the code in `add_cflags` and `get_libs` identical in how they deal with dependencies (which makes total sense)

**Comment by ypujante** [comment]
After merging with the latest master, all the tests about sizes are gone (thank you @sbc100 for the commit).

The only failure is this one:

```text
ERROR [61.331s]: test_offset_converter (test_browser.browser64)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/project/test/common.py", line 2040, in run_browser
    self.assertContained(expected, output)
  File "/root/project/test/common.py", line 1370, in assertContained
    additional_info
  File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
    raise self.failureException(msg)
AssertionError: Expected to find '/report_result?exit:1
' in '[no http server activity]
', diff:

--- expected
+++ actual
@@ -1 +1 @@
-/report_result?exit:1
+[no http server activity]
```

which seems to be another spurious test failure and not related to my changes.

**Comment by sbc100** [comment]
> After merging with the latest master, all the tests about sizes are gone (thank you @sbc100 for the commit).
> 
> The only failure is this one:
> 
> ```
> ERROR [61.331s]: test_offset_converter (test_browser.browser64)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/root/project/test/common.py", line 2040, in run_browser
>     self.assertContained(expected, output)
>   File "/root/project/test/common.py", line 1370, in assertContained
>     additional_info
>   File "/usr/lib/python3.6/unittest/case.py", line 670, in fail
>     raise self.failureException(msg)
> AssertionError: Expected to find '/report_result?exit:1
> ' in '[no http server activity]
> ', diff:
> 
> --- expected
> +++ actual
> @@ -1 +1 @@
> -/report_result?exit:1
> +[no http server activity]
> ```
> 
> which seems to be another spurious test failure and not related to my changes.

Yes this type of flake is not uncommon in the browser tests.  I restarted them, after which this should land.

**Comment by ypujante** [comment]
@sbc100 Thank you for restarting the failed test which now works. So it can be merged if you approve it.

**Review by sbc100 — APPROVED** [review]

**Pull request merged**
